src/server.rs

```diff
@@ -200,40 +290,31 @@ impl Server {
         }
     }
 
-    /// Sends header of `FramebufferUpdate` message containing number of rectangles to be sent.
+    /// Checks validity of passed data and sends `FramebufferUpdate` message.
     ///
-    /// Call to this method must be followed by `count` calls to `send_rectangle_header`.
-    pub fn send_framebuffer_update_header(&mut self, count: u16) -> Result<()> {
-        try!(protocol::S2C::FramebufferUpdate{count}.write_to(&mut self.stream));
-        Ok(())
-    }
+    /// This is all-or-nothing method. Rectangles will be sent only if all rectangles are valid.
+    pub fn send_framebuffer_update(&mut self, rectangles: &[Rectangle]) -> Result<()> {
+        // Check validity
+        for rectangle in rectangles {
+            try!(rectangle.check(self.bytes_per_pixel));
+        }
 
-    /// Sends rectangle header.
-    ///
-    /// The rectangle header must be followed by the pixel data in the specified encoding.
-    pub fn send_rectangle_header(&mut self,
-                                 x: u16,
-                                 y: u16,
-                                 width: u16,
-                                 height: u16,
-                                 encoding: protocol::Encoding)
-                                 -> Result<()> {
-        try!(protocol::Rectangle {
-            x_position: x,
-            y_position: y,
-            width: width,
-            height: height,
-            encoding: encoding,
-        }.write_to(&mut self.stream));
+        // Send the message
+        let count = rectangles.len() as u16;
+        try!(protocol::S2C::FramebufferUpdate{count}.write_to(&mut self.stream));
+        for rectangle in rectangles {
+            try!(rectangle.write_to(&mut self.stream));
+        }
         Ok(())
     }
 
-    /// Writes raw data to the socket.
-    ///
-    /// This method may be used to send pixel data following rectangle header sent by
-    /// `send_rectangle_header` method.
-    pub fn send_raw_data(&mut self, data: &[u8]) -> Result<()> {
-        try!(self.stream.write_all(data));
+    /// Sends `FramebufferUpdate` message without checking validity of its contents.
+    pub fn send_framebuffer_update_unchecked(&mut self, rectangles: &[Rectangle]) -> Result<()> {
```

Comment:
Do we need this? The check is not expensive at all, especially compared to cost of I/O. This just clutters the API.